sqlbase: remove ColumnType from EncDatum

RaduBerinde opened the pull request:
#### distsqlrun: require types for RowBuffer

Curretly the RowBuffer (used mainly in tests) relies on the types in the
`EncDatum`s (unless it is an empty buffer). To prepare for a removing the type
information from `EncDatum`, requires the types in all cases.

#### sqlbase: remove ColumnType from EncDatum

We carry around the entire `ColumnType` with every `EncDatum`. Initially
`ColumnType` was just a 32-bit value (it used memory that is otherwise lost to
padding). But in time this type grew much bigger.

This change modifies the EncDatum interface to require the type information to
be passed in. We leave the type as a deprecated field and assert that the type
matches. In a subsequent commit, this type will be removed.

In addition to the immediate performance improvement, these changes should help
facilitate switching to a more efficient representation of rows or row groups
(which would definitely require the type information to be passed in).

#### sqlbase: remove the deprecated type from EncDatum

Infrastructure benchmarks for this group of changes:

```
name                        old time/op    new time/op    delta
Infrastructure/n1/r1-4         186µs ± 4%     187µs ±10%     ~       (p=0.959 n=8+8)
Infrastructure/n1/r100-4       395µs ±29%     345µs ±12%  -12.51%    (p=0.019 n=9+9)
Infrastructure/n1/r10000-4    16.9ms ±24%    14.1ms ±44%     ~     (p=0.052 n=10+10)
Infrastructure/n3/r1-4         598µs ± 2%     580µs ± 5%   -3.02%   (p=0.043 n=7+10)
Infrastructure/n3/r100-4      1.30ms ± 6%    1.15ms ± 3%  -12.09%   (p=0.000 n=9+10)
Infrastructure/n3/r10000-4    56.9ms ±14%    45.6ms ± 4%  -19.79%   (p=0.000 n=10+9)

name                        old alloc/op   new alloc/op   delta
Infrastructure/n1/r1-4        37.3kB ± 0%    33.0kB ± 0%  -11.63%   (p=0.000 n=10+9)
Infrastructure/n1/r100-4       187kB ± 0%     123kB ± 0%  -34.51%  (p=0.000 n=10+10)
Infrastructure/n1/r10000-4    15.8MB ± 0%     9.6MB ± 0%  -39.25%  (p=0.000 n=10+10)
Infrastructure/n3/r1-4         129kB ± 2%     117kB ± 2%   -9.18%  (p=0.000 n=10+10)
Infrastructure/n3/r100-4       725kB ± 1%     489kB ± 1%  -32.50%  (p=0.000 n=10+10)
Infrastructure/n3/r10000-4    61.3MB ± 0%    39.1MB ± 0%  -36.16%  (p=0.000 n=10+10)

name                        old allocs/op  new allocs/op  delta
Infrastructure/n1/r1-4           214 ± 0%       217 ± 0%   +1.59%   (p=0.000 n=8+10)
Infrastructure/n1/r100-4         326 ± 0%       328 ± 0%   +0.61%    (p=0.002 n=6+6)
Infrastructure/n1/r10000-4     10.2k ± 0%     10.1k ± 0%   -0.69%   (p=0.000 n=10+9)
Infrastructure/n3/r1-4           935 ± 0%       937 ± 0%   +0.26%   (p=0.000 n=10+8)
Infrastructure/n3/r100-4       1.71k ± 0%     1.71k ± 0%     ~     (p=0.346 n=10+10)
Infrastructure/n3/r10000-4     76.6k ± 0%     76.4k ± 0%   -0.19%  (p=0.000 n=10+10)
```

CC @knz

RaduBerinde committed:
sqlbase: remove ColumnType from EncDatum

We carry around the entire `ColumnType` with every `EncDatum`. Initially
`ColumnType` was just a 32-bit value (it used memory that is otherwise lost to
padding). But in time this type grew much bigger.

This change modifies the EncDatum interface to require the type information to
be passed in. We leave the type as a deprecated field and assert that the type
matches. In a subsequent commit, this type will be removed.

In addition to the immediate performance improvement, these changes should help
facilitate switching to a more efficient representation of rows or row groups
(which would definitely require the type information to be passed in).

RaduBerinde committed:
distsqlrun: require types for RowBuffer

Curretly the RowBuffer (used mainly in tests) relies on the types in the
`EncDatum`s (unless it is an empty buffer). To prepare for a removing the type
information from `EncDatum`, requires the types in all cases.

RaduBerinde committed:
sqlbase: remove the deprecated type from EncDatum

Infrastructure benchmarks for this group of changes:

```
name                        old time/op    new time/op    delta
Infrastructure/n1/r1-4         186µs ± 4%     187µs ±10%     ~       (p=0.959 n=8+8)
Infrastructure/n1/r100-4       395µs ±29%     345µs ±12%  -12.51%    (p=0.019 n=9+9)
Infrastructure/n1/r10000-4    16.9ms ±24%    14.1ms ±44%     ~     (p=0.052 n=10+10)
Infrastructure/n3/r1-4         598µs ± 2%     580µs ± 5%   -3.02%   (p=0.043 n=7+10)
Infrastructure/n3/r100-4      1.30ms ± 6%    1.15ms ± 3%  -12.09%   (p=0.000 n=9+10)
Infrastructure/n3/r10000-4    56.9ms ±14%    45.6ms ± 4%  -19.79%   (p=0.000 n=10+9)

name                        old alloc/op   new alloc/op   delta
Infrastructure/n1/r1-4        37.3kB ± 0%    33.0kB ± 0%  -11.63%   (p=0.000 n=10+9)
Infrastructure/n1/r100-4       187kB ± 0%     123kB ± 0%  -34.51%  (p=0.000 n=10+10)
Infrastructure/n1/r10000-4    15.8MB ± 0%     9.6MB ± 0%  -39.25%  (p=0.000 n=10+10)
Infrastructure/n3/r1-4         129kB ± 2%     117kB ± 2%   -9.18%  (p=0.000 n=10+10)
Infrastructure/n3/r100-4       725kB ± 1%     489kB ± 1%  -32.50%  (p=0.000 n=10+10)
Infrastructure/n3/r10000-4    61.3MB ± 0%    39.1MB ± 0%  -36.16%  (p=0.000 n=10+10)

name                        old allocs/op  new allocs/op  delta
Infrastructure/n1/r1-4           214 ± 0%       217 ± 0%   +1.59%   (p=0.000 n=8+10)
Infrastructure/n1/r100-4         326 ± 0%       328 ± 0%   +0.61%    (p=0.002 n=6+6)
Infrastructure/n1/r10000-4     10.2k ± 0%     10.1k ± 0%   -0.69%   (p=0.000 n=10+9)
Infrastructure/n3/r1-4           935 ± 0%       937 ± 0%   +0.26%   (p=0.000 n=10+8)
Infrastructure/n3/r100-4       1.71k ± 0%     1.71k ± 0%     ~     (p=0.346 n=10+10)
Infrastructure/n3/r10000-4     76.6k ± 0%     76.4k ± 0%   -0.19%  (p=0.000 n=10+10)
```

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19246)
<!-- Reviewable:end -->

asubiotto commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 11 of 11 files at r1, 49 of 54 files at r2, 2 of 2 files at r3.
Review status: 50 of 55 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/algebraic_set_op_test.go, line 83 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19246#-Kw_nAQ3aglcoZeItvss:-Kw_nAQ3aglcoZeItvst:bv1kmat) ([raw file](https://github.com/cockroachdb/cockroach/blob/a38d30bcd1c31ff99bd3d0ca71033d249923ced4/pkg/sql/distsqlrun/algebraic_set_op_test.go#L83)):*
> ```Go
> 
> func runProcessors(tc testCase) (sqlbase.EncDatumRows, error) {
> 	types := []sqlbase.ColumnType{intType, intType}
> ```

Could use `twoIntCols` here.

---

*[pkg/sql/distsqlrun/mergejoiner_test.go, line 42 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19246#-Kwa0eb626tvMuA9NLWp:-Kwa0eb626tvMuA9NLWq:b-ayegej) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68ed89b085c340f095e1bd399c5b152f2170c6c/pkg/sql/distsqlrun/mergejoiner_test.go#L42)):*
> ```Go
> 		spec       MergeJoinerSpec
> 		outCols    []uint32
> 		leftTypes  []sqlbase.ColumnType
> ```

Consider using this method of creating types in `sorter_test` as well. It would be nice not to have different tests use different ways to create types. However, I also don't like that you would have to create the types manually and `sorter_test` has longer int rows than the global test variables you've introduced. Maybe you could create some function that takes a length and returns an intType slice of that length.

---

*[pkg/sql/sqlbase/table.go, line 428 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19246#-Kw_zgvS3RZ8hNI85RjD:-Kw_zgvS3RZ8hNI85RjE:b-zapi5j) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68ed89b085c340f095e1bd399c5b152f2170c6c/pkg/sql/sqlbase/table.go#L428)):*
> ```Go
> 	if len(values) != len(types) {
> 		panic(fmt.Sprintf("%d values, %d types", len(values), len(types)))
> 		//return nil, errors.Errorf("%d values, %d types", len(values), len(types))
> ```

Delete this line?

---

*[pkg/util/encoding/encoding.go, line 567 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19246#-Kwa-6LTgtgp-PcT76Db:-Kwa-6LTgtgp-PcT76Dc:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68ed89b085c340f095e1bd399c5b152f2170c6c/pkg/util/encoding/encoding.go#L567)):*
> ```Go
> 		if i == -1 {
> 			panic(fmt.Sprintf("did not find terminator %#x in buffer %#x", e.escape, b))
> 			//return 0, errors.Errorf("did not find terminator %#x in buffer %#x", e.escape, b)
> ```

ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19246#-:-Kwa214iqjy5CtsTRbVA:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Updated. Also removed the unnecessary `alloc` param from `DecodeIndexKey`(CC @richardwu)

---

Review status: 44 of 54 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/algebraic_set_op_test.go, line 83 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19246#-Kw_nAQ3aglcoZeItvss:-KwatzPYp6M3FHOoIGM2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a38d30bcd1c31ff99bd3d0ca71033d249923ced4/pkg/sql/distsqlrun/algebraic_set_op_test.go#L83)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Could use `twoIntCols` here.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/mergejoiner_test.go, line 42 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19246#-Kwa0eb626tvMuA9NLWp:-KwaxluBOXDDBcQZs6NA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68ed89b085c340f095e1bd399c5b152f2170c6c/pkg/sql/distsqlrun/mergejoiner_test.go#L42)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Consider using this method of creating types in `sorter_test` as well. It would be nice not to have different tests use different ways to create types. However, I also don't like that you would have to create the types manually and `sorter_test` has longer int rows than the global test variables you've introduced. Maybe you could create some function that takes a length and returns an intType slice of that length.
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/table.go, line 428 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19246#-Kw_zgvS3RZ8hNI85RjD:-Kwatzv7AEoFlzN12R0W:beit15y) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68ed89b085c340f095e1bd399c5b152f2170c6c/pkg/sql/sqlbase/table.go#L428)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Delete this line?
</blockquote></details>

Oops, this panic was used to debug some issues, reverted.

---

*[pkg/util/encoding/encoding.go, line 567 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19246#-Kwa-6LTgtgp-PcT76Db:-Kwau4kf6_6Zjx3wYAW1:bvth56w) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68ed89b085c340f095e1bd399c5b152f2170c6c/pkg/util/encoding/encoding.go#L567)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

reverted

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19246#-:-KwayNRvOmtozx0ZvoN5:b-eal0tp)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: 1 of 54 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/distsqlrun/algebraic_set_op_test.go, line 83 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/19246#-KwepJrfjV3BiFiyZ293:-KwepJrfjV3BiFiyZ294:bd334g2) ([raw file](https://github.com/cockroachdb/cockroach/blob/28fecfdc7b42f99ea7fd8fc6a635c0db71fa11c4/pkg/sql/distsqlrun/algebraic_set_op_test.go#L83)):*
> ```Go
> 
> func runProcessors(tc testCase) (sqlbase.EncDatumRows, error) {
> 	types := twoIntCols
> ```

`twoIntCols` could be passed directly to `NewRowBuffer` here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19246)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/distsqlrun/algebraic_set_op_test.go, line 83 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/19246#-KwepJrfjV3BiFiyZ293:-Kwevt-6uBj6UYF4u3iQ:bfssli) ([raw file](https://github.com/cockroachdb/cockroach/blob/28fecfdc7b42f99ea7fd8fc6a635c0db71fa11c4/pkg/sql/distsqlrun/algebraic_set_op_test.go#L83)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

`twoIntCols` could be passed directly to `NewRowBuffer` here.
</blockquote></details>

it's used below too

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19246)*
<!-- Sent from Reviewable.io -->
